defyx/doc/configuration.md
# DefyX configuration

DefyX has 45 customizable parameters (see table below). We recommend each project using DefyX to select a unique configuration to prevent network attacks from hashpower rental services.

These parameters can be modified in source file [configuration.h](../src/configuration.h).

|parameter|description|default value|
|---------|-----|-------|
|`RANDOMX_ARGON_MEMORY`|The number of 1 KiB Argon2 blocks in the Cache| `262144`|
|`RANDOMX_ARGON_ITERATIONS`|The number of Argon2d iterations for Cache initialization|`3`|
|`RANDOMX_ARGON_LANES`|The number of parallel lanes for Cache initialization|`1`|
|`RANDOMX_ARGON_SALT`|Argon2 salt|`"DefyX\x03"`|
|`RANDOMX_CACHE_ACCESSES`|The number of random Cache accesses per Dataset item|`8`|
|`RANDOMX_SUPERSCALAR_LATENCY`|Target latency for SuperscalarHash (in cycles of the reference CPU)|`170`|
|`RANDOMX_DATASET_BASE_SIZE`|Dataset base size in bytes|`2147483648`|
|`RANDOMX_DATASET_EXTRA_SIZE`|Dataset extra size in bytes|`33554368`|
|`RANDOMX_PROGRAM_SIZE`|The number of instructions in a DefyX program|`256`|
|`RANDOMX_PROGRAM_ITERATIONS`|The number of iterations per program|`2048`|
|`RANDOMX_PROGRAM_COUNT`|The number of programs per hash|`8`|
|`RANDOMX_JUMP_BITS`|Jump condition mask size in bits|`8`|
|`RANDOMX_JUMP_OFFSET`|Jump condition mask offset in bits|`8`|
|`RANDOMX_SCRATCHPAD_L3`|Scratchpad size in bytes|`2097152`|
|`RANDOMX_SCRATCHPAD_L2`|Scratchpad L2 size in bytes|`262144`|
|`RANDOMX_SCRATCHPAD_L1`|Scratchpad L1 size in bytes|`16384`|
|`RANDOMX_FREQ_*` (29x)|Instruction frequencies|multiple values|

Not all of the parameters can be changed safely and most parameters have some contraints on what values can be selected (checked at compile-time).

**Disclaimer: The compile-time checks only prevent obviously broken configurations. Passing the checks does not imply that the configuration is safe and will not cause crashes or other issues. We recommend that each non-standard configuration is thoroughly tested before being deployed.**

### RANDOMX_ARGON_MEMORY

This parameter determines the amount of memory needed in the light mode. Memory is specified in KiB (1 KiB = 1024 bytes).

#### Permitted values
Integer powers of 2 in the range 1 - 2097152.

#### Notes
Lower sizes will reduce the memory-hardness of the algorithm.

### RANDOMX_ARGON_ITERATIONS

Determines the number of passes of Argon2 that are used to generate the Cache.

#### Permitted values
Any positive integer.

#### Notes
The time needed to initialize the Cache is proportional to the value of this constant.

### RANDOMX_ARGON_LANES

The number of parallel lanes for Cache initialization.

#### Permitted values
Any positive integer.

#### Notes
This parameter determines how many threads can be used for Cache initialization. 

### RANDOMX_ARGON_SALT

Salt value for Cache initialization.

#### Permitted values
Any string of byte values.

#### Note
Every implementation should choose a unique salt value.

### RANDOMX_CACHE_ACCESSES

The number of random Cache access per Dataset item.

#### Permitted values
Any integer greater than 1.

#### Notes
This value directly determines the performance ratio between the 'fast' and 'light' modes. 

### RANDOMX_SUPERSCALAR_LATENCY
Target latency for SuperscalarHash, in cycles of the reference CPU.

#### Permitted values
Integers in the range 1 - 10000.

#### Notes
The default value was tuned so that a high-performance superscalar CPU running at 2-4 GHz will execute SuperscalarHash in similar time it takes to load data from RAM (40-80 ns). Using a lower value will make Dataset generation (and light mode) more memory bound, while increasing this value will make Dataset generation (and light mode) more compute bound.

### RANDOMX_DATASET_BASE_SIZE

Dataset base size in bytes.

#### Permitted values
Integer powers of 2 in the range 64 - 4294967296 (inclusive).

#### Note
This constant affects the memory requirements in fast mode. Some values are unsafe depending on other parameters. See [Unsafe configurations](#unsafe-configurations).

### RANDOMX_DATASET_EXTRA_SIZE

Dataset extra size in bytes.

#### Permitted values
Non-negative integer divisible by 64.

#### Note
This constant affects the memory requirements in fast mode. Some values are unsafe depending on other parameters. See [Unsafe configurations](#unsafe-configurations).

### RANDOMX_PROGRAM_SIZE

The number of instructions in a DefyX program.

#### Permitted values
Positive integers divisible by 8 in the range 8 - 32768 (inclusive).

#### Notes
Smaller values will make DefyX more DRAM-latency bound, while higher values will make DefyX more compute-bound. Some values are unsafe. See [Unsafe configurations](#unsafe-configurations).

### RANDOMX_PROGRAM_ITERATIONS

The number of iterations per program.

#### Permitted values
Any positive integer.

#### Notes
Time per hash increases linearly with this constant. Smaller values will increase the overhead of program compilation, while larger values may allow more time for optimizations. Some values are unsafe. See [Unsafe configurations](#unsafe-configurations).

### RANDOMX_PROGRAM_COUNT

The number of programs per hash.

#### Permitted values
Any positive integer.

#### Notes
Time per hash increases linearly with this constant. Some values are unsafe. See [Unsafe configurations](#unsafe-configurations).

### RANDOMX_JUMP_BITS
Jump condition mask size in bits.

#### Permitted values
Positive integers. The sum of `RANDOMX_JUMP_BITS` and `RANDOMX_JUMP_OFFSET` must not exceed 16.

#### Notes
This determines the jump probability of the CBRANCH instruction. The default value of 8 results in jump probability of <code>1/2<sup>8</sup> = 1/256</code>. Increasing this constant will decrease the rate of jumps (and vice versa).

### RANDOMX_JUMP_OFFSET
Jump condition mask offset in bits.

#### Permitted values
Non-negative integers. The sum of `RANDOMX_JUMP_BITS` and `RANDOMX_JUMP_OFFSET` must not exceed 16.

#### Notes
Since the low-order bits of DefyX registers are slightly biased, this offset moves the condition mask to higher bits, which are less biased. Using values smaller than the default may result in a slightly lower jump probability than the theoretical value calculated from `RANDOMX_JUMP_BITS`.

### RANDOMX_SCRATCHPAD_L3
DefyX Scratchpad size in bytes.

#### Permitted values
Any integer power of 2. Must be larger than or equal to `RANDOMX_SCRATCHPAD_L2`.

#### Notes

The default value of 2 MiB was selected to match the typical cache/core ratio of desktop processors. Using a lower value will make DefyX more core-bound, while using larger values will make the algorithm more latency-bound. Some values are unsafe depending on other parameters. See [Unsafe configurations](#unsafe-configurations).

### RANDOMX_SCRATCHPAD_L2

Scratchpad L2 size in bytes.

#### Permitted values
Any integer power of 2. Must be larger than or equal to `RANDOMX_SCRATCHPAD_L1`.

#### Notes
The default value of 256 KiB was selected to match the typical per-core L2 cache size of desktop processors. Using a lower value will make DefyX more core-bound, while using larger values will make the algorithm more latency-bound.

### RANDOMX_SCRATCHPAD_L1

Scratchpad L1 size in bytes.

#### Permitted values
Any integer power of 2. The minimum is 64 bytes.

#### Notes
The default value of 16 KiB was selected to be about half of the per-core L1 cache size of desktop processors. Using a lower value will make DefyX more core-bound, while using larger values will make the algorithm more latency-bound.

### RANDOMX_FREQ_*

Instruction frequencies (per 256 instructions).

#### Permitted values
There is a total of 29 different instructions. The sum of frequencies must be equal to 256.

#### Notes

Making changes to the default values is not recommended. The only exceptions are the instruction pairs IROR_R/IROL_R, FADD_R/FSUB_R and FADD_M/FSUB_M, which are functionally equivalent. Example of a safe custom configuration:

||default|custom|
|-|------|------|-|
|`RANDOMX_FREQ_IROR_R`|8|5|
|`RANDOMX_FREQ_IROL_R`|2|5|

||default|custom|
|-|------|------|
|`RANDOMX_FREQ_FADD_R`|16|17|
|`RANDOMX_FREQ_FSUB_R`|16|15|

||default|custom|
|-|------|------|
|`RANDOMX_FREQ_FADD_M`|5|4|
|`RANDOMX_FREQ_FSUB_M`|5|6|

## Unsafe configurations

There are some configurations that are considered 'unsafe' because they affect the security of the algorithm against attacks. If the conditions listed below are not satisfied, the configuration is unsafe and a compilation error is emitted when building the DefyX library.

These checks can be disabled by definining `RANDOMX_UNSAFE` when building DefyX, e.g. by using `-DRANDOMX_UNSAFE` command line switch in GCC or MSVC. It is not recommended to disable these checks except for testing purposes.


### 1. Memory-time tradeoffs

#### Condition
```` 
RANDOMX_CACHE_ACCESSES * RANDOMX_ARGON_MEMORY * 1024 + 33554432 >= RANDOMX_DATASET_BASE_SIZE + RANDOMX_DATASET_EXTRA_SIZE
```` 

Configurations not satisfying this condition are vulnerable to memory-time tradeoffs, which enables efficient mining in light mode.

#### Solutions

* Increase `RANDOMX_CACHE_ACCESSES` or `RANDOMX_ARGON_MEMORY`. 
* Decrease `RANDOMX_DATASET_BASE_SIZE` or `RANDOMX_DATASET_EXTRA_SIZE`.

### 2. Insufficient Scratchpad writes

#### Condition
```` 
(128 + RANDOMX_PROGRAM_SIZE * RANDOMX_FREQ_ISTORE / 256) * (RANDOMX_PROGRAM_COUNT * RANDOMX_PROGRAM_ITERATIONS) >= RANDOMX_SCRATCHPAD_L3
```` 

Configurations not satisfying this condition are vulnerable to Scratchpad size optimizations due to low amount of writes.

#### Solutions

* Increase `RANDOMX_PROGRAM_SIZE`, `RANDOMX_FREQ_ISTORE`, `RANDOMX_PROGRAM_COUNT` or `RANDOMX_PROGRAM_ITERATIONS`. 
* Decrease `RANDOMX_SCRATCHPAD_L3`.

### 3. Program filtering strategies

#### Condition
```
RANDOMX_PROGRAM_COUNT > 1
```

Configurations not satisfying this condition are vulnerable to program filtering strategies.

#### Solution

* Increase `RANDOMX_PROGRAM_COUNT` to at least 2.

### 4. Low program entropy

#### Condition
```
RANDOMX_PROGRAM_SIZE >= 64
```

Configurations not satisfying this condition do not have a sufficient number of instruction combinations. 

#### Solution

* Increase `RANDOMX_PROGRAM_SIZE` to at least 64.

### 5. High compilation overhead

#### Condition
```
RANDOMX_PROGRAM_ITERATIONS >= 400
```

Configurations not satisfying this condition have a program compilation overhead exceeding 10%.

#### Solution

* Increase `RANDOMX_PROGRAM_ITERATIONS` to at least 400.



CHANGELOG.md
# v2.16.0-beta
- [#1036](https://github.com/xlarig/xlarig/pull/1036) Added RandomWOW (RandomX with different preferences) algorithm support for [Wownero](http://wownero.org/).
  - Algorithm name used by miner is `randomx/wow` or `rx/wow`.
  - Currently runtime algorithm switching NOT supported with other algorithms.

# v2.15.4-beta
- Added global uptime and extended connection information in API.
- API now return current algorithm instead of global algorithm specified in config.
- This version also include all changes from stable version v2.14.4.

# v2.15.3-beta
- [#1014](https://github.com/xlarig/xlarig/issues/1014) Fixed regression, default value for `algo` option was not applied.

# v2.15.2-beta
- [#1010](https://github.com/xlarig/xlarig/pull/1010#issuecomment-482632107) Added daemon support (solo mining).
- [#1012](https://github.com/xlarig/xlarig/pull/1012) Fixed compatibility with clang 9.
- Config subsystem was rewritten, internally JSON is primary format now.
- Fixed regression, big HTTP responses was truncated.

# v2.15.1-beta
- [#1007](https://github.com/xlarig/xlarig/issues/1007) Old HTTP API backend based on libmicrohttpd, replaced to custom HTTP server (libuv + http_parser).
- [#257](https://github.com/xlarig/xlarig-nvidia/pull/257) New logging subsystem, file and syslog now always without colors.

# v2.15.0-beta
- [#314](https://github.com/xlarig/xlarig-proxy/issues/314) Added donate over proxy feature.
  - Added new option `donate-over-proxy`.
  - Added real graceful exit.
  
# v2.14.4
- [#992](https://github.com/xlarig/xlarig/pull/992)  Fixed compilation with Clang 3.5.
- [#1012](https://github.com/xlarig/xlarig/pull/1012) Fixed compilation with Clang 9.0.
- In HTTP API for unknown hashrate now used `null` instead of `0.0`.
- Fixed MSVC 2019 version detection.
- Removed obsolete automatic variants.

# v2.14.1
* [#975](https://github.com/xlarig/xlarig/issues/975) Fixed crash on Linux if double thread mode used.

# v2.14.0
- **[#969](https://github.com/xlarig/xlarig/pull/969) Added new algorithm `cryptonight/rwz`, short alias `cn/rwz` (also known as CryptoNight ReverseWaltz), for upcoming [Graft](https://www.graft.network/) fork.**
- **[#931](https://github.com/xlarig/xlarig/issues/931) Added new algorithm `cryptonight/zls`, short alias `cn/zls` for [Zelerius Network](https://zelerius.org) fork.**
- **[#940](https://github.com/xlarig/xlarig/issues/940) Added new algorithm `cryptonight/double`, short alias `cn/double` (also known as CryptoNight HeavyX), for [X-CASH](https://x-cash.org/).**
- [#951](https://github.com/xlarig/xlarig/issues/951#issuecomment-469581529) Fixed crash if AVX was disabled on OS level.
- [#952](https://github.com/xlarig/xlarig/issues/952) Fixed compile error on some Linux.
- [#957](https://github.com/xlarig/xlarig/issues/957#issuecomment-468890667) Added support for embedded config.
- [#958](https://github.com/xlarig/xlarig/pull/958) Fixed incorrect user agent on ARM platforms.
- [#968](https://github.com/xlarig/xlarig/pull/968) Optimized `cn/r` algorithm performance.

# v2.13.1
- [#946](https://github.com/xlarig/xlarig/pull/946) Optimized software AES implementations for CPUs without hardware AES support. `cn/r`, `cn/wow` up to 2.6 times faster, 4-9% improvements for other algorithms.

# v2.13.0
- **[#938](https://github.com/xlarig/xlarig/issues/938) Added support for new algorithm `cryptonight/r`, short alias `cn/r` (also known as CryptoNightR or CryptoNight variant 4), for upcoming [Monero](https://www.getmonero.org/) fork on March 9, thanks [@SChernykh](https://github.com/SChernykh).**
- [#939](https://github.com/xlarig/xlarig/issues/939) Added support for dynamic (runtime) pools reload.
- [#932](https://github.com/xlarig/xlarig/issues/932) Fixed `cn-pico` hashrate drop, regression since v2.11.0.

# v2.12.0
- [#929](https://github.com/xlarig/xlarig/pull/929) Added support for new algorithm `cryptonight/wow`, short alias `cn/wow` (also known as CryptonightR), for upcoming [Wownero](http://wownero.org) fork on February 14.

# v2.11.0
- [#928](https://github.com/xlarig/xlarig/issues/928) Added support for new algorithm `cryptonight/gpu`, short alias `cn/gpu` (original name `cryptonight-gpu`), for upcoming [Ryo currency](https://ryo-currency.com) fork on February 14.
- [#749](https://github.com/xlarig/xlarig/issues/749) Added support for detect hardware AES in runtime on ARMv8 platforms.
- [#292](https://github.com/xlarig/xlarig/issues/292) Fixed build on ARMv8 platforms if compiler not support hardware AES.

# v2.10.0
- [#904](https://github.com/xlarig/xlarig/issues/904) Added new algorithm `cn-pico/trtl` (aliases `cryptonight-turtle`, `cn-trtl`) for upcoming TurtleCoin (TRTL) fork.
- Default value for option `max-cpu-usage` changed to `100` also this option now deprecated.

# v2.9.4
- [#913](https://github.com/xlarig/xlarig/issues/913) Fixed Masari (MSR) support (this update required for upcoming fork).
- [#915](https://github.com/xlarig/xlarig/pull/915) Improved security, JIT memory now read-only after patching.

# v2.9.3
- [#909](https://github.com/xlarig/xlarig/issues/909) Fixed compile errors on FreeBSD.
- [#912](https://github.com/xlarig/xlarig/pull/912) Fixed, C++ implementation of `cn/half` was produce up to 13% of invalid hashes.

# v2.9.2
- [#907](https://github.com/xlarig/xlarig/pull/907) Fixed crash on Linux.

# v2.9.1
- Restored compatibility with https://stellite.hashvault.pro.

# v2.9.0
- [#899](https://github.com/xlarig/xlarig/issues/899) Added support for new algorithm `cn/half` for Masari and Stellite forks.
- [#834](https://github.com/xlarig/xlarig/pull/834) Added ASM optimized code for AMD Bulldozer.
- [#839](https://github.com/xlarig/xlarig/issues/839) Fixed FreeBSD compile.
- [#857](https://github.com/xlarig/xlarig/pull/857) Fixed impossible to build for macOS without clang.

# v2.8.3
- [#813](https://github.com/xlarig/xlarig/issues/813) Fixed critical bug with Minergate pool and variant 2.

# v2.8.1

defyx/src/configuration.h
/*
Copyright (c) 2018-2019, tevador <tevador@gmail.com>

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
	* Redistributions of source code must retain the above copyright
	  notice, this list of conditions and the following disclaimer.
	* Redistributions in binary form must reproduce the above copyright
	  notice, this list of conditions and the following disclaimer in the
	  documentation and/or other materials provided with the distribution.
	* Neither the name of the copyright holder nor the
	  names of its contributors may be used to endorse or promote products
	  derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#pragma once

//Cache size in KiB. Must be a power of 2.
#define RANDOMX_ARGON_MEMORY	   131072

//Number of Argon2d iterations for Cache initialization.
#define RANDOMX_ARGON_ITERATIONS   2

//Number of parallel lanes for Cache initialization.
#define RANDOMX_ARGON_LANES        1

//Argon2d salt
#define RANDOMX_ARGON_SALT         "DefyXScala\x13"

//Number of random Cache accesses per Dataset item. Minimum is 2.
#define RANDOMX_CACHE_ACCESSES     2

//Target latency for SuperscalarHash (in cycles of the reference CPU).
#define RANDOMX_SUPERSCALAR_LATENCY   170

//Dataset base size in bytes. Must be a power of 2.
#define RANDOMX_DATASET_BASE_SIZE  33554432

//Dataset extra size. Must be divisible by 64.
#define RANDOMX_DATASET_EXTRA_SIZE 33554368

//Number of instructions in a DefyX program. Must be divisible by 8.
#define RANDOMX_PROGRAM_SIZE       64

//Number of iterations during VM execution.
#define RANDOMX_PROGRAM_ITERATIONS 1024

//Number of chained VM executions per hash.
#define RANDOMX_PROGRAM_COUNT      4

//Scratchpad L3 size in bytes. Must be a power of 2.
#define RANDOMX_SCRATCHPAD_L3      262144

//Scratchpad L2 size in bytes. Must be a power of two and less than or equal to RANDOMX_SCRATCHPAD_L3.
#define RANDOMX_SCRATCHPAD_L2      131072

//Scratchpad L1 size in bytes. Must be a power of two (minimum 64) and less than or equal to RANDOMX_SCRATCHPAD_L2.
#define RANDOMX_SCRATCHPAD_L1      65536

//Jump condition mask size in bits.
#define RANDOMX_JUMP_BITS          8

//Jump condition mask offset in bits. The sum of RANDOMX_JUMP_BITS and RANDOMX_JUMP_OFFSET must not exceed 16.
#define RANDOMX_JUMP_OFFSET        8

/*
Instruction frequencies (per 256 opcodes)
Total sum of frequencies must be 256
*/

//Integer instructions
#define RANDOMX_FREQ_IADD_RS       25
#define RANDOMX_FREQ_IADD_M         7
#define RANDOMX_FREQ_ISUB_R        16
#define RANDOMX_FREQ_ISUB_M         7
#define RANDOMX_FREQ_IMUL_R        16
#define RANDOMX_FREQ_IMUL_M         4
#define RANDOMX_FREQ_IMULH_R        4
#define RANDOMX_FREQ_IMULH_M        1
#define RANDOMX_FREQ_ISMULH_R       4
#define RANDOMX_FREQ_ISMULH_M       1
#define RANDOMX_FREQ_IMUL_RCP       8
#define RANDOMX_FREQ_INEG_R         2
#define RANDOMX_FREQ_IXOR_R        15
#define RANDOMX_FREQ_IXOR_M         5
#define RANDOMX_FREQ_IROR_R         8
#define RANDOMX_FREQ_IROL_R         2
#define RANDOMX_FREQ_ISWAP_R        4

//Floating point instructions
#define RANDOMX_FREQ_FSWAP_R        4
#define RANDOMX_FREQ_FADD_R        16
#define RANDOMX_FREQ_FADD_M         5
#define RANDOMX_FREQ_FSUB_R        16
#define RANDOMX_FREQ_FSUB_M         5
#define RANDOMX_FREQ_FSCAL_R        6
#define RANDOMX_FREQ_FMUL_R        32
#define RANDOMX_FREQ_FDIV_M         4
#define RANDOMX_FREQ_FSQRT_R        6

//Control instructions
#define RANDOMX_FREQ_CBRANCH       16
#define RANDOMX_FREQ_CFROUND        1

//Store instruction
#define RANDOMX_FREQ_ISTORE        16

//No-op instruction
#define RANDOMX_FREQ_NOP            0
/*                               ------
                                  256
*/
